File: promise/src/test/java/com/voxeet/testpromise/sub_promise/promise/PromiseTestWithSumDelayed2.java
package com.voxeet.testpromise.sub_promise.promise;

import com.voxeet.promise.Promise;
import com.voxeet.promise.solve.PromiseSolver;
import com.voxeet.testpromise.utils.AndroidMockUtil;

import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

/**
 * Created by kevinleperf on 06/04/2018.
 */

public class PromiseTestWithSumDelayed2 {

    @Before
    public void setHandler() {
        Promise.setHandler(AndroidMockUtil.mockMainThreadHandler());
    }

    @Test
    public void test() throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);
        final int[] final_result = {0};

        System.out.println("executing test");
        new Promise<>((PromiseSolver<Integer>) solver -> solver.resolve(new Promise<>(solver1 -> new Thread() {
            @Override
            public void run() {
                System.out.println("sleeping...");
                try {
                    sleep(20);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                solver1.resolve(10 * 10);
            }
        }.start())))
                .then((result, solver) -> final_result[0] = result)
                .error(error -> {
                    System.out.println("error catched");
                    final_result[0] = 0;
                    error.printStackTrace();
                    latch.countDown();
                });

        //6s are enough
        latch.await(6, TimeUnit.SECONDS);
        if (final_result[0] != 100) {
            throw new IllegalStateException("Expected 100... having " + final_result[0]);
        } else {
            System.out.println("having result " + final_result[0]);
        }
    }
}

File: promise/src/test/kotlin/com/voxeet/testpromise/subpromise/promise/PromiseTestWithSumDelayed2.kt
package com.voxeet.testpromise.subpromise.promise

import com.voxeet.promise.Promise
import com.voxeet.promise.solve.Solver
import com.voxeet.promise.solve.ThenVoid
import com.voxeet.testpromise.mockedhandler
import org.junit.Before
import org.junit.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class PromiseTestWithSumDelayed2 {
    @Before
    fun setHandler() {
        Promise.setHandler(mockedhandler())
    }

    @Test
    @Throws(InterruptedException::class)
    fun test() {
        val latch = CountDownLatch(1)
        val final_result = intArrayOf(0)
        println("executing test")
        Promise { solver: Solver<Int?> ->
            solver.resolve(
                Promise { solver1: Solver<Int?> ->
                    object : Thread() {
                        override fun run() {
                            println("sleeping...")
                            try {
                                sleep(20)
                            } catch (e: InterruptedException) {
                                e.printStackTrace()
                            }
                            solver1.resolve(10 * 10)
                        }
                    }.start()
                }
            )
        }.then(ThenVoid<Int?> { result: Int? ->
            final_result[0] = result!!
        }).error { error: Throwable ->
            println("error catched")
            final_result[0] = 0
            error.printStackTrace()
            latch.countDown()
        }

        //6s are enough
        latch.await(6, TimeUnit.SECONDS)
        check(final_result[0] == 100) { "Expected 100... having " + final_result[0] }
        println("having result " + final_result[0])
    }
}
